resources/aws_ec2.go

```diff
@@ -0,0 +1,535 @@
+						stateName = *stateOutput.Reservations[0].Instances[0].State.Name
+					}
+					if len(stateOutput.Reservations) == 0 || (len(stateOutput.Reservations) == 1 && stateName != "stopped") {
+						x := &chanStruct{
+							str: "running",
+						}
+						select {
+						case awsChan <- x:
+						case <-closeChan:
+							return
+						}
+					}
+				}
+			}
+			if obj.State == "terminated" {
+				obj.client.WaitUntilInstanceExistsWithContext(ctx, diInput)
+				stateOutput, err := obj.client.DescribeInstances(diInput)
+				if err != nil {
+					awsChan <- &chanStruct{
+						err: errwrap.Wrapf(err, "error describing instances"),
+					}
+				}
+				if len(stateOutput.Reservations) == 1 {
+					{
+						x := &chanStruct{
+							str: "exists",
+						}
+						select {
+						case awsChan <- x:
+						case <-closeChan:
+							return
+						}
+					}
+				}
+			}
+		}
+	}()
+	for {
+		select {
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit
+			}
+		case msg, ok := <-awsChan:
+			if !ok {
+				return *exit
+			}
+			if err := msg.err; err != nil {
+				return err
+			}
+			log.Println("instance", msg.str)
+			obj.StateOK(false)
+			send = true
+		}
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// CheckApply method for AwsEc2 resource.
+func (obj *AwsEc2Res) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.prependName())},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending"), aws.String("stopped"), aws.String("stopping")},
+			},
+		},
+	}
+	diOutput, err := obj.client.DescribeInstances(&diInput)
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error describing instances")
+	}
+
+	if len(diOutput.Reservations) < 1 && obj.State == "terminated" {
+		return true, nil
+	}
+	if len(diOutput.Reservations) == 1 && *diOutput.Reservations[0].Instances[0].State.Name == obj.State {
+		return true, nil
+	}
+	if !apply {
+		return false, nil
+	}
+
+	if len(diOutput.Reservations) > 1 {
+		return false, fmt.Errorf("too many reservations")
+	}
+	if len(diOutput.Reservations) == 1 {
+		instanceID := diOutput.Reservations[0].Instances[0].InstanceId
+		describeInput := &ec2.DescribeInstancesInput{
+			InstanceIds: []*string{instanceID},
+		}
+		if len(diOutput.Reservations[0].Instances) > 1 {
+			return false, fmt.Errorf("more than one instance was returned")
+		}
+		if obj.State == "running" {
+			startInput := &ec2.StartInstancesInput{
+				InstanceIds: []*string{instanceID},
+			}
+			_, err := obj.client.StartInstances(startInput)
+			if err != nil {
+				return false, errwrap.Wrapf(err, "error starting instance")
+			}
+			obj.client.WaitUntilInstanceRunning(describeInput)
```

Comment:
do these waits have a guaranteed timeout, and do they always return eventually??
eg: what if an aws region blows up while we're running it? Can we use the ctx version with a specific timeout that we choose and set as a constant, or if you think it's a good idea, as a struct option?